### ISSUE TITLE
$product->get_type() returning 'variable' and the condition check is wrong

### ISSUE THREAD

**Issue by rahul-acodez** [woothemes/woocommerce#32753]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Hi,
We found a bug,
In the file \woocommerce\includes\class-wc-ajax.php
Line number 411

if ( $product && 'variation' === $product->get_type() ) {
//codes
}

Here, $product->get_type() returning 'variable' But the function expects 'variation'. The condition check is wrong in this case.

Please check the issue immediately. We can't get the variation  id inside the filter 'woocommerce_add_cart_item_data'

I saw the same condition check is used so many places in your code. I don't how it changed from the original.

### Expected behavior

In the mentioned line the condition check should be true,

it should be like,

if ( $product && 'variable' === $product->get_type() ) {
//codes
}

### Actual behavior

Here, In this case we are receiving variation_id always 0

### Steps to reproduce

![unnamed](https://user-images.githubusercontent.com/56103755/165029065-8bfb7194-a56f-4595-98e9-a2df9a2b1794.png)

Here, you can see the type and the condition check both are mismatching.

Error occurring, If you manually create an ajax functionality in product detail page

Because in listing page woocommerce doesn't support ajax add to cart for variation products

### WordPress Environment

<html>
<body>

WordPress address (URL): |   | http://localhost/wcpa-pro
-- | -- | --
Site address (URL): |   | http://localhost/wcpa-pro
WooCommerce version: |   | 6.4.1
WooCommerce REST API package: |   | 6.4.1 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\includes
WooCommerce Blocks package: |   | 7.2.2 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\woocommerce-blocks/
Action Scheduler package: |   | 3.4.0 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\action-scheduler
WooCommerce Admin package: |   | 3.3.2 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\woocommerce-admin
Log directory writable: |   | D:\Plugins\Custom Product Addon\Pro/wp-content/uploads/wc-logs/
WordPress version: |   | 5.9.3
WordPress multisite: |   | –
WordPress memory limit: |   | 512 MB
WordPress debug mode: |   |  
WordPress cron: |   |  
Language: |   | en_US
External object cache: |   | –

</body>
</html>WordPress address (URL):		http://localhost/wcpa-pro
Site address (URL):		http://localhost/wcpa-pro
WooCommerce version:		6.4.1
WooCommerce REST API package:		 6.4.1 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\includes
WooCommerce Blocks package:		 7.2.2 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\woocommerce-blocks/
Action Scheduler package:		 3.4.0 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\action-scheduler
WooCommerce Admin package:		 3.3.2 D:\Plugins\Custom Product Addon\Pro\wp-content\plugins\woocommerce\packages\woocommerce-admin
Log directory writable:		 D:\Plugins\Custom Product Addon\Pro/wp-content/uploads/wc-logs/
WordPress version:		5.9.3
WordPress multisite:		–
WordPress memory limit:		512 MB
WordPress debug mode:		
WordPress cron:		
Language:		en_US
External object cache:		–
Server environment
Server info:		Apache/2.4.41 (Win64) OpenSSL/1.1.1c PHP/7.4.1
PHP version:		7.4.1
PHP post max size:		40 MB
PHP time limit:		0
PHP max input vars:		1000
cURL version:		7.67.0, OpenSSL/1.1.1c
SUHOSIN installed:		–
MySQL version:		5.5.5-10.4.18-MariaDB
Max upload size:		40 MB
Default timezone is UTC:		
fsockopen/cURL:		
SoapClient:		 Your server does not have the [SoapClient](https://php.net/manual/en/class.soapclient.php) class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument:		
GZip:		
Multibyte string:		
Remote post:		
Remote get:		
Database
WooCommerce database version:		6.3.1
Database prefix	 	wcpa_
Total Database Size	 	8.63MB
Database Data Size	 	5.74MB
Database Index Size	 	2.89MB
wcpa_woocommerce_sessions	 	Data: 0.08MB + Index: 0.02MB + Engine InnoDB
wcpa_woocommerce_api_keys	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_woocommerce_attribute_taxonomies	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_woocommerce_downloadable_product_permissions	 	Data: 0.02MB + Index: 0.17MB + Engine InnoDB
wcpa_woocommerce_order_items	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_woocommerce_order_itemmeta	 	Data: 0.06MB + Index: 0.05MB + Engine InnoDB
wcpa_woocommerce_tax_rates	 	Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wcpa_woocommerce_tax_rate_locations	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_woocommerce_shipping_zones	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_woocommerce_shipping_zone_locations	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_woocommerce_shipping_zone_methods	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_woocommerce_payment_tokens	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_woocommerce_payment_tokenmeta	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_woocommerce_log	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_actionscheduler_actions	 	Data: 0.08MB + Index: 0.13MB + Engine InnoDB
wcpa_actionscheduler_claims	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_actionscheduler_groups	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_actionscheduler_logs	 	Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wcpa_as3cf_items	 	Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wcpa_commentmeta	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_comments	 	Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wcpa_infinite_uploads_files	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_links	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_mailchimp_carts	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_mailchimp_jobs	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_mcloud_task	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_mcloud_task_data	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_mcloud_task_schedule	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_mcloud_task_token	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_options	 	Data: 4.05MB + Index: 0.08MB + Engine InnoDB
wcpa_postmeta	 	Data: 0.19MB + Index: 0.13MB + Engine InnoDB
wcpa_posts	 	Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wcpa_termmeta	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_terms	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_term_relationships	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_term_taxonomy	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_usermeta	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_users	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_wc_admin_notes	 	Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wcpa_wc_admin_note_actions	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_bookings_availability	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_wc_bookings_availabilitymeta	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_booking_relationships	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_category_lookup	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_wc_customer_lookup	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_wc_download_log	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_wc_order_composite_lookup	 	Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wcpa_wc_order_coupon_lookup	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_order_product_lookup	 	Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wcpa_wc_order_stats	 	Data: 0.02MB + Index: 0.31MB + Engine InnoDB
wcpa_wc_order_tax_lookup	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_product_attributes_lookup	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wc_product_meta_lookup	 	Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wcpa_wc_rate_limits	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_wc_reserved_stock	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_wc_tax_rate_classes	 	Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wcpa_wc_warranty_products	 	Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wcpa_wc_webhooks	 	Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wcpa_wdp_orders	 	Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wcpa_wdp_order_items	 	Data: 0.02MB + Index: 0.16MB + Engine InnoDB
wcpa_wdp_persistent_rules_cache	 	Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wcpa_wdp_rules	 	Data: 0.02MB + Index: 0.06MB + Engine InnoDB
Post Type Counts
attachment	 	24
csp_rules	 	1
page	 	9
post	 	2
product	 	21
product_variation	 	7
revision	 	1
shop_coupon	 	1
shop_order	 	13
warranty_request	 	1
wc_booking	 	4
wcpa_pt_forms	 	2
Security
Secure connection (HTTPS):		 Your store is not using HTTPS. [Learn more about HTTPS and SSL Certificates](https://docs.woocommerce.com/document/ssl-and-https/).
Hide errors from visitors		Error messages should not be shown to visitors.
Active plugins (4)
[Woocommerce Custom Product Addons](https://acowebs.com/)	 	by Acowebs – 4.1.8
Advanced Dynamic Pricing for WooCommerce	 	by AlgolPlus – 4.1.0
[WOOCS - WooCommerce Currency Switcher](https://currency-switcher.com/)	 	by realmag777 – 2.3.6.2
[WooCommerce](https://woocommerce.com/)	 	by Automattic – 6.4.1
Inactive plugins (15)
[Akismet Anti-Spam](https://akismet.com/)	 	by Automattic – 4.2.2
[Checkout Field Editor for WooCommerce](https://www.themehigh.com/)	 	by ThemeHigh – 1.6.0
[Hello Dolly](http://wordpress.org/plugins/hello-dolly/)	 	by Matt Mullenweg – 1.7.2
Infinite Uploads	 	by Infinite Uploads – 1.1.5
[Mailchimp for WooCommerce](https://mailchimp.com/connect-your-store/)	 	by Mailchimp – 2.6.2
[Media Cloud](https://github.com/interfacelab/ilab-media-tools)	 	by interfacelab – 4.4.0
[Product Feed PRO for WooCommerce](https://www.adtribes.io/support/?utm_source=wpadmin&utm_medium=plugin&utm_campaign=woosea_product_feed_pro)	 	by AdTribes.io – 11.5.3
[Role Based Pricing for WooCommerce](https://woocommerce.com/products/role-based-pricing-for-woocommerce/)	 	by Addify – 1.5.6
[WooCommerce Bookings](https://woocommerce.com/products/woocommerce-bookings/)	 	by WooCommerce – 1.15.48
[WooCommerce PayPal Payments](https://woocommerce.com/products/woocommerce-paypal-payments/)	 	by WooCommerce – 1.7.1
[WooCommerce Warranty Requests](https://woocommerce.com/products/warranty-requests/)	 	by WooCommerce – 1.9.30
[WordPress Importer](https://wordpress.org/plugins/wordpress-importer/)	 	by wordpressdotorg – 0.7
[WP Hide & Security Enhancer](https://wp-hide.com/)	 	by Nsp Code – 1.7.8
[WP Offload Media Lite](http://wordpress.org/extend/plugins/amazon-s3-and-cloudfront/)	 	by Delicious Brains – 2.6.2
[YITH WooCommerce Gift Cards Premium](https://yithemes.com/themes/plugins/yith-woocommerce-gift-cards)	 	by YITH – 3.7.0
Settings
API enabled:		–
Force SSL:		–
Currency		INR (₹)
Currency position		left
Thousand separator		,
Decimal separator		.
Number of decimals		2
Taxonomies: Product types		booking (booking), composite (composite), external (external), gift-card (gift-card), grouped (grouped), simple (simple), variable (variable)
Taxonomies: Product visibility		exclude-from-catalog (exclude-from-catalog), exclude-from-search (exclude-from-search), featured (featured), outofstock (outofstock), rated-1 (rated-1), rated-2 (rated-2), rated-3 (rated-3), rated-4 (rated-4), rated-5 (rated-5)
Connected to WooCommerce.com:		–
WooCommerce pages
[Shop base](http://localhost/wcpa-pro/wp-admin/post.php?post=6&action=edit):		#6 - /shop/
[Cart](http://localhost/wcpa-pro/wp-admin/post.php?post=7&action=edit):		#7 - /cart/
[Checkout](http://localhost/wcpa-pro/wp-admin/post.php?post=8&action=edit):		#8 - /checkout/
[My account](http://localhost/wcpa-pro/wp-admin/post.php?post=9&action=edit):		#9 - /my-account/
Terms and conditions:		 Page not set
Theme
Name:		Storefront
Version:		4.0.0 (update to version 4.1.0 is available)
Author URL:		https://woocommerce.com/
Child theme:		 – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: [How to create a child theme](https://developer.wordpress.org/themes/advanced-topics/child-themes/)
WooCommerce support:		
Templates
Overrides:	 	–
Action Scheduler
Version:	3.4.0
Data store:	ActionScheduler_DBStore
Action Status	 	Count	Oldest Scheduled Date	Newest Scheduled Date
Complete	 	150	2022-03-30 06:58:36 +0000	2022-04-25 05:37:58 +0000
Pending	 	1	2022-05-11 04:04:39 +0000	2022-05-11 04:04:39 +0000
Status report information
Generated at:	 	2022-04-25 06:07:40 +00:00

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by woocommercebot** [comment]
Hi @rahul-acodez,

  Thank you for opening the issue! It requires further feedback from the WooCommerce Core team.

  We are adding the `needs developer feedback` label to this issue so that the Core team could take a look.

  Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by jorgeatorres** [comment]
Hi @rahul-acodez!

Thanks for the report. Would you mind sharing steps to replicate the problem? Also, a minimal snippet showing how you're implementing the AJAX functionality you've mentioned can help us diagnose this better.

After a quick glance at the code in `class-wc-ajax.php`, we believe it is correct. Under normal circumstances, the "Add to Cart" button which this AJAX is tied to isn't shown for variable products but, in any case, if the product isn't a 'variation' but a 'variable' one, it wouldn't make much sense to try to obtain its parent, which is precisely what this code section does.

There might be something going on with the interaction between your custom code and this AJAX handler that we would like to see.

**Comment by lkraav** [comment]
To those being confused here with me, there are indeed 2 separate, but similarly named product types

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-product-variable.php
https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-product-variation.php

"Variable" is a parent product.
"Variation" is a child of "variable" product.
